File: ComplexityOFSorting.cpp
#include <iostream>
#include <math.h>
#include <chrono>
#include <unistd.h>

using namespace std;

// Merge Sort 
void merge(int arr[], int low, int mid, int up)
{
	int lenL = mid - low + 1;
	int lenR = up - mid;

	int L[lenL], R[lenR];

	for (int i = 0; i < lenL; i++)
		L[i] = arr[low + i];
	for (int j = 0; j < lenR; j++)
		R[j] = arr[mid + 1 + j];


	int i = 0;

	int j = 0;

	int k = low;

	while (i < lenL && j < lenR) {
		if (L[i] <= R[j]) {
			arr[k] = L[i];
			i++;
		}
		else {
			arr[k] = R[j];
			j++;
		}
		k++;
	}

	while (i < lenL) {
		arr[k] = L[i];
		i++;
		k++;
	}

	while (j < lenR) {
		arr[k] = R[j];
		j++;
		k++;
	}
}
void mergeSort(int arr[],int low,int up){
	if(up<=low)
		return;
	int mid =(up+low)/2;
	mergeSort(arr,low,mid);
	mergeSort(arr,mid+1,up);
	merge(arr,low,mid,up);
}

// Insertion Sort
void insertionSort(int arr[], int n)
{
	int i, key, j;
	for (i = 1; i < n; i++)
	{
		key = arr[i];
		j = i - 1;

		while (j >= 0 && arr[j] > key)
		{
			arr[j + 1] = arr[j];
			j = j - 1;
		}
		arr[j + 1] = key;
	}
}

// Bitonic Sort
void compAndSwap(int a[], int i, int j, int dir)
{
	if (dir==(a[i]>a[j]))
		swap(a[i],a[j]);
}
void bitonicMerge(int a[], int low, int cnt, int dir)
{
	if (cnt>1)
	{
		int k = cnt/2;
		for (int i=low; i<low+k; i++)
			compAndSwap(a, i, i+k, dir);
		bitonicMerge(a, low, k, dir);
		bitonicMerge(a, low+k, k, dir);
	}
}
void bitonicSort(int a[],int low, int cnt, int dir)
{
	if (cnt>1)
	{
		int k = cnt/2;
		bitonicSort(a, low, k, 1);
		bitonicSort(a, low+k, k, 0);
		bitonicMerge(a,low, cnt, dir);
	}
}

// Bubble Sort
void Nonstringbubble_sort(int  arr[], int length)
{
	bool flag = true;
	for (size_t i = 0; i < length - 1 && flag; i++)
	{
		flag = false;
		for (size_t j = 0; j < length - i - 1; j++)
		{
			if (arr[j] > arr[j + 1])
			{
				swap(arr[j], arr[j + 1]);
				flag = true;
			}
		}
	}
}

// Selection Sort
void selectionSort(int arr[], int n)
{
	int i, j, min_idx;

	for (i = 0; i < n-1; i++) 
    { 
        min_idx = i; 
        for (j = i+1; j < n; j++) 
        if (arr[j] < arr[min_idx]) 
            min_idx = j; 
  
        swap(arr[min_idx], arr[i]); 
    }
}

// Shell Sort
int shellSort(int arr[], int n)
{
	for (int gap = n/2; gap > 0; gap /= 2)
    {
        
        for (int i = gap; i < n; i += 1)
        {
            int temp = arr[i];

           	int j = i - gap;
			while (j >= 0  && arr[j] > temp)
			{
				arr[j + gap] = arr[j];
				j = j - gap;
			}
			arr[j + gap] = temp;
        }
    }
	return 0;
}

template<class T>
void CopyTo(T source , T Destination , int sourceLen)
{
	for (size_t i = 0; i < sourceLen; i++)
	{
		Destination[i] = source[i];
	}
	
}

void printArray(int A[], int size , string title)
{
	cout << title << " : ";
	for (int i = 0; i < size; i++)
		cout << A[i] << " ";
	cout << endl;
}

int main()
{
	const int arraylen = pow(2,12);
	int arr[arraylen] ;
	for (size_t i = 0; i < arraylen; i++)
		arr[i] = rand()%10+10;

	int arrMerge[arraylen] ;
	int arrInsertion[arraylen] ;
	int arrBubble[arraylen] ;
	int arrSelection[arraylen] ;
	int arrShell[arraylen] ;
	int arrBitonic[arraylen] ;

	CopyTo(arr, arrMerge, arraylen);
	CopyTo(arr, arrInsertion, arraylen);
	CopyTo(arr, arrBubble, arraylen);
	CopyTo(arr, arrSelection, arraylen);
	CopyTo(arr, arrShell, arraylen);
	CopyTo(arr, arrBitonic, arraylen);


	auto start = chrono::steady_clock::now();
	mergeSort(arrMerge,0,arraylen-1);
	auto end = chrono::steady_clock::now();
	auto MergeTimens= chrono::duration_cast<chrono::nanoseconds>(end - start).count();

	start = chrono::steady_clock::now();
	insertionSort(arrInsertion,arraylen);
	end = chrono::steady_clock::now();
	auto InsertionTimens= chrono::duration_cast<chrono::nanoseconds>(end - start).count();

	start = chrono::steady_clock::now();
	bitonicSort(arrBitonic,0,arraylen,1);
	end = chrono::steady_clock::now();
	auto bitonicTimens= chrono::duration_cast<chrono::nanoseconds>(end - start).count();

	start = chrono::steady_clock::now();
	Nonstringbubble_sort(arrBubble,arraylen);
	end = chrono::steady_clock::now();
	auto BubbleTimens= chrono::duration_cast<chrono::nanoseconds>(end - start).count();

	start = chrono::steady_clock::now();
	selectionSort(arrSelection,arraylen);
	end = chrono::steady_clock::now();
	auto SelectionTimens= chrono::duration_cast<chrono::nanoseconds>(end - start).count();

	start = chrono::steady_clock::now();
	shellSort(arrShell,arraylen);
	end = chrono::steady_clock::now();
	auto ShellTimens= chrono::duration_cast<chrono::nanoseconds>(end - start).count();
  	


	// printArray(arr ,arraylen,          "UnSorted Array       " );
	// printArray(arrMerge ,arraylen,     "Sorted whit Merge    " );
	// printArray(arrInsertion ,arraylen, "Sorted whit Insertion" );
	// printArray(arrBubble ,arraylen,    "Sorted whit Bubble   " );
	// printArray(arrSelection ,arraylen, "Sorted whit Selection" );
	// printArray(arrShell ,arraylen,     "Sorted whit Shell    " );
	// printArray(arrBitonic ,arraylen,   "Sorted whit Bitonic  " );

	
	cout << ">> Merge Time Elapsed     : " << MergeTimens << " ns" << endl;
	cout << ">> Insertion Time Elapsed : " << InsertionTimens << " ns" << endl;
	cout << ">> Bitonic Time Elapsed   : " << bitonicTimens << " ns" << endl;
	cout << ">> Bubble Time Elapsed    : " << BubbleTimens << " ns" << endl;
	cout << ">> Selection Time Elapsed : " << SelectionTimens << " ns" << endl;
	cout << ">> Shell Time Elapsed     : " << ShellTimens << " ns" << endl;

	return 0;
}

File: ComplexityOFSorting.cs
using System;
enum Direction
{
	descending,
	ascending
}
class SortClass
{
	static void merge(int[] arr, int low, int mid, int up, Direction dir)
	{

		int lenL = mid - low + 1;
		int lenR = up - mid;

		int[] L = new int[lenL];
		int[] R = new int[lenR];

		int i = 0;
		int j = 0;
		int k = low;

		Array.Copy(arr , low , L , 0 , lenL);
		Array.Copy(arr , mid + 1 ,  R , 0 , lenR);

		if(dir == Direction.ascending)
        {
			while (i < lenL && j < lenR)
			{
				if (L[i] <= R[j])
					arr[k++] = L[i++];
				else
					arr[k++] = R[j++];
			}
		}
        else
        {
			while (i < lenL && j < lenR)
			{
				if (L[i] > R[j])
					arr[k++] = L[i++];
				else
					arr[k++] = R[j++];
			}
		}

		while (i < lenL)
			arr[k++] = L[i++];

		while (j < lenR)
			arr[k++] = R[j++];
	}
	static void sort(int[] arr, int low, int up , Direction dir)
	{
		if (low < up)
		{
			int mid = (low + up) / 2;

			sort(arr, low, mid , dir);
			sort(arr, mid + 1, up , dir);

			merge(arr, low, mid, up , dir);
		}
	}
	static void Swap<T>(ref T lhs, ref T rhs)
	{
		T temp;
		temp = lhs;
		lhs = rhs;
		rhs = temp;
	}
	public static void compAndSwap(int[] a, int i, int j, Direction dir)
	{
		if ((a[i] > a[j]))
		{
			if (dir == Direction.ascending)
			{
				Swap(ref a[i], ref a[j]);
			}
		}
		else if (dir == Direction.descending)
		{
			Swap(ref a[i], ref a[j]);
		}

	}
	public static void bitonicMerge(int[] a, int low, int count, Direction dir)
	{
		if (count > 1)
		{
			int k = count / 2;
			for (int i = low; i < low + k; i++)
				compAndSwap(a, i, i + k, dir);

			bitonicMerge(a, low, k, dir);
			bitonicMerge(a, low + k, k, dir);
		}
	}
	public static void bitonicSort(int[] a, int low, int count, Direction dir)
	{
		if (count > 1)
		{
			int k = count / 2;

			bitonicSort(a, low, k, Direction.ascending);

			bitonicSort(a, low + k, k, Direction.descending);

			bitonicMerge(a, low, count, dir);
		}
	}
	public static void Insertionsort(int[] arr, Direction dir)
	{
		for (int i = 1; i < arr.Length; ++i)
		{
			int key = arr[i];
			int j = i - 1;

			if (dir == Direction.ascending)
			{
				while (j >= 0 && arr[j] > key)
				{
					arr[j + 1] = arr[j];
					j = j - 1;
				}
			}
			else
			{
				while (j >= 0 && arr[j] < key)
				{
					arr[j + 1] = arr[j];
					j = j - 1;
				}
			}

			arr[j + 1] = key;
		}
	}
	static void Selectionsort(int[] arr, Direction dir)
	{
		int n = arr.Length;
		bool flag = false;
		for (int i = 0; i < n - 1; i++)
		{
			int min_idx = i;
				if(dir == Direction.ascending)
                {
				for (int j = i + 1; j < n; j++)
					if (arr[j] < arr[min_idx])
					{
						min_idx = j;
						flag = true;
					}
				}
				else
                {
				for (int j = i + 1; j < n; j++)
					if (arr[j] > arr[min_idx])
					{
						min_idx = j;
						flag = true;
					}
			}

			if (flag)
			{
				int temp = arr[min_idx];
				arr[min_idx] = arr[i];
				arr[i] = temp;
			}
		}
	}
	static void Shellsort(int[] arr, Direction dir)
	{
		int n = arr.Length;

		for (int gap = n / 2; gap > 0; gap /= 2)
		{

			for (int i = gap; i < n; i += 1)
			{
				
				int temp = arr[i];

				int j;
				for (j = i; j >= gap && arr[j - gap] > temp; j -= gap)
					arr[j] = arr[j - gap];

				
				arr[j] = temp;
			}
		}

	}
	static void bubbleSort(int[] arr, Direction dir)
	{
		int n = arr.Length;
		bool flag = true;
		for (int i = 0; i < n - 1 && flag; i++)
        {
			flag = false;
			if(dir == Direction.ascending)
            {
				for (int j = 0; j < n - i - 1; j++)
					if (arr[j] > arr[j + 1])
					{
						int temp = arr[j];
						arr[j] = arr[j + 1];
						arr[j + 1] = temp;
						flag = true;
					}
			}
            else 
			{
				for (int j = 0; j < n - i - 1; j++)
					if (arr[j] < arr[j + 1])
					{
						int temp = arr[j];
						arr[j] = arr[j + 1];
						arr[j + 1] = temp;
						flag = true;
					}
			}
		}
	}

	public static void PrintArray(int[] arr , string Title)
    {
		Console.Write(Title + " : ");
		foreach (int tar in arr)
        {
			Console.Write(tar);
			Console.Write(" ");
		}
		Console.WriteLine();
	}
	public static void Main()
	{
		var watch_bitonicSort = new System.Diagnostics.Stopwatch();
		var watch_Insertionsort = new System.Diagnostics.Stopwatch();
		var watch_Selectionsort = new System.Diagnostics.Stopwatch();
		var watch_Shellsort = new System.Diagnostics.Stopwatch();
		var watch_bubbleSort = new System.Diagnostics.Stopwatch();
		var watch_MergeSort = new System.Diagnostics.Stopwatch();

		int arrayLen = (int)Math.Pow(2, 16);
		int[] array = new int[arrayLen];

		Random rd = new Random();
		for(int i = 0; i < arrayLen; i++)
        {
			array[i] = rd.Next(1, 1000);
		}

		int[] array_bitonicSort = new int[arrayLen];
		var array_Insertionsort = new int[arrayLen];
		var array_Selectionsort = new int[arrayLen];
		var array_Shellsort = new int[arrayLen];
		var array_bubbleSort = new int[arrayLen];
		var array_MergeSort = new int[arrayLen];

		Array.Copy(array, array_bitonicSort, arrayLen);
		Array.Copy(array, array_Insertionsort, arrayLen);
		Array.Copy(array, array_Selectionsort, arrayLen);
		Array.Copy(array, array_Shellsort, arrayLen);
		Array.Copy(array, array_bubbleSort, arrayLen);
		Array.Copy(array, array_MergeSort, arrayLen);


		watch_bitonicSort.Start();
		bitonicSort(array_bitonicSort, 0, array_bitonicSort.Length, Direction.ascending);
		watch_bitonicSort.Stop();

		watch_Insertionsort.Start();
		Insertionsort(array_Insertionsort, Direction.ascending);
		watch_Insertionsort.Stop();

		watch_Selectionsort.Start();
		Selectionsort(array_Selectionsort, Direction.ascending);
		watch_Selectionsort.Stop();

		watch_Shellsort.Start();
		Shellsort(array_Shellsort, Direction.ascending);
		watch_Shellsort.Stop();

		watch_bubbleSort.Start();
		bubbleSort(array_bubbleSort, Direction.ascending);
		watch_bubbleSort.Stop();

		watch_MergeSort.Start();
		sort(array_MergeSort, 0, arrayLen-1, Direction.ascending);
		watch_MergeSort.Stop();

		//PrintArray(array,               "UnSorted Array     ");
		//PrintArray(array_bitonicSort,   "bitonicSort Array  ");
		//PrintArray(array_Insertionsort, "Insertionsort Array");
		//PrintArray(array_Selectionsort, "Selectionsort Array");
		//PrintArray(array_Shellsort,     "Shellsort Array    ");
		//PrintArray(array_bubbleSort,    "bubbleSort Array   ");
		//PrintArray(array_MergeSort,    "mergeSort Array   ");
		//Console.WriteLine();
		//Console.WriteLine();
		//Console.WriteLine();
		//Console.WriteLine();
		//Console.WriteLine();
		Console.WriteLine("> Time Elapsed : ");
		Console.WriteLine(">> Bitonic Sort Time Elapsed :   {0}", watch_bitonicSort.Elapsed);
		Console.WriteLine(">> Insertion Sort Time Elapsed : {0}", watch_Insertionsort.Elapsed);
		Console.WriteLine(">> Shell Sort Time Elapsed :     {0}", watch_Shellsort.Elapsed);
		Console.WriteLine(">> Selection Sort Time Elapsed : {0}", watch_Selectionsort.Elapsed);
		Console.WriteLine(">> Bubble Sort Time Elapsed :    {0}", watch_bubbleSort.Elapsed);
		Console.WriteLine(">> Merge Sort Time Elapsed :   {0}", watch_MergeSort.Elapsed);

		Console.ReadLine();
	}
}

